pennylane/labs/resource_estimation/templates/stateprep.py

```diff
@@ -0,0 +1,595 @@
+            inplace controlled semi-adder operation (see figure 4. of
+            `arXiv:2409.07332 <https://arxiv.org/pdf/2409.07332>`)_.
+
+        Returns:
+            list[GateCount]: A list of GateCount objects, where each object
+            represents a specific quantum gate and the number of times it appears
+            in the decomposition.
+        """
+        gate_counts = []
+        precision = precision or kwargs["config"]["precision_qrom_state_prep"]
+
+        expected_size = num_state_qubits if positive_and_real else num_state_qubits + 1
+        if isinstance(selswap_depths, int) or selswap_depths is None:
+            selswap_depths = [selswap_depths] * expected_size
+
+        num_precision_wires = math.ceil(math.log2(math.pi / precision)) + 1
+        gate_counts.append(AllocWires(num_precision_wires))
+
+        for j in range(num_state_qubits):
+            num_bitstrings = 2**j
+            num_bit_flips = max(2 ** (j - 1), 1)
+
+            gate_counts.append(
+                GateCount(
+                    re.ResourceQROM.resource_rep(
+                        num_bitstrings=num_bitstrings,
+                        size_bitstring=num_precision_wires,
+                        num_bit_flips=num_bit_flips,
+                        clean=False,
+                        select_swap_depth=selswap_depths[j],
+                    )
+                )
+            )
+
+            gate_counts.append(
+                GateCount(
+                    re.ResourceAdjoint.resource_rep(
+                        re.resource_rep(
+                            re.ResourceQROM,
+                            {
+                                "num_bitstrings": num_bitstrings,
+                                "num_bit_flips": num_bit_flips,
+                                "size_bitstring": num_precision_wires,
+                                "clean": False,
+                                "select_swap_depth": selswap_depths[j],
+                            },
+                        ),
+                    )
+                )
+            )
+
+        if use_phase_grad_trick:
+            semi_adder = re.ResourceSemiAdder.resource_rep(max_register_size=num_precision_wires)
+            h = re.ResourceHadamard.resource_rep()
+            s = re.ResourceS.resource_rep()
+            s_dagg = re.ResourceAdjoint.resource_rep(base_cmpr_op=s)
+            gate_counts.append(
+                GateCount(
+                    re.ResourceControlled.resource_rep(
+                        base_cmpr_op=semi_adder, num_ctrl_wires=1, num_ctrl_values=0
+                    ),
+                    count=num_state_qubits,
+                )
+            )
+            gate_counts.append(GateCount(h, 2 * num_precision_wires))
+            gate_counts.append(GateCount(s, num_precision_wires))
+            gate_counts.append(
+                GateCount(s_dagg, num_precision_wires)
+            )  # map RY rotations to RZ for phase grad
```

Comment:
Should the phase gradient cost be included here? Maybe it is better if we assume we have it 
Imagine you are using QSVT and QROMStatePrep for the block encoding. You would be creating a new phase gradient per iteration

Reply:
This is just doing the Semi-adder it's not including the cost of creating the phase gradient register itself right.

Reply:
Oh I see now 👍